roboteam_observer/observer/src/filters/vision/ball/CameraGroundBallFilter.cpp

```diff
@@ -1,54 +1,134 @@
+    rtt::Angle robotYaw = rtt::Angle(robot.position.yaw);
+    rtt::Vector2 ballPos = rtt::Vector2(positionEstimate.x(), positionEstimate.y());
+    rtt::Vector2 ballVel = rtt::Vector2(velocityEstimate.x(), velocityEstimate.y());
+    rtt::Vector2 robotVel = rtt::Vector2(robot.velocity.velocity.x(), robot.velocity.velocity.y());
+
+    if (!rtt::RobotShape(robotPos, 0.07, 0.09, robotYaw).contains(ballPos)) {
+        return false;
+    }
+
+    auto ballVelUsed = ballVel - robotVel;
+    auto robotShape = rtt::RobotShape(robotPos, 0.07 + 0.0213, 0.09 + 0.0213, robotYaw);
+    auto frontLine = robotShape.kicker();
+    auto ballVelLine = rtt::LineSegment(ballPos, ballPos - ballVelUsed.stretchToLength(0.09 * 5));
+
+    if (frontLine.intersects(ballVelLine)) {
+        auto collisionAngle = (robotYaw - rtt::Angle(ballVelUsed.scale(-1)));
+
+        if (std::abs(collisionAngle) > 0.5 * M_PI) {
+            return false;
+        }
+
+        auto outVelAbs = ballVelUsed.length() - (ballVelUsed.length() * std::cos(collisionAngle) * 1);
+        auto outVel = rtt::Vector2(robotYaw).scale(-1).rotate(collisionAngle).scale(-1).stretchToLength(outVelAbs);
+        outVel += robotVel;
+        Eigen::Vector2d outVelEigen(outVel.x, outVel.y);
+        ekf.setVelocity(outVelEigen);
+        return true;
+    }
+
+    auto botCircle = rtt::Circle(robotPos, 0.09 + 0.0213);
+    auto intersects = botCircle.intersects(ballVelLine);
+
+    if (!intersects.empty()) {
+        auto intersect = intersects[0];
+        auto collisionAngle = (rtt::Vector2(robotPos - intersect).toAngle() - ballVelUsed.toAngle());
+
+        if (std::abs(collisionAngle) > 0.5 * M_PI) {
+            return false;
+        }
+
+        auto outVelAbs = ballVelUsed.length() - (ballVelUsed.length() * std::cos(collisionAngle) * 0.6);
+        auto outVel = rtt::Vector2(robotPos - intersect).rotate(collisionAngle).scale(-1).stretchToLength(outVelAbs);
+        outVel += robotVel;
+        Eigen::Vector2d outVelEigen(outVel.x, outVel.y);
+        ekf.setVelocity(outVelEigen);
+        return true;
+    }
+
+    return false;
+}
+
+bool CameraGroundBallFilter::checkRobots(const std::vector<FilteredRobot>& robots, const Eigen::Vector2d& positionEstimate, const Eigen::Vector2d& velocityEstimate) {
+    for (const auto& robot : robots) {
+        if (checkRobotCollision(robot, positionEstimate, velocityEstimate)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+CameraGroundBallPrediction CameraGroundBallFilter::predict(Time time, std::vector<FilteredRobot> yellowRobots, std::vector<FilteredRobot> blueRobots) {
+    auto positionEstimate = ekf.getPositionEstimate(time);
+    auto velocityEstimate = ekf.getVelocityEstimate(time);
+
+    if (!checkRobots(yellowRobots, positionEstimate, velocityEstimate)) {
+        checkRobots(blueRobots, positionEstimate, velocityEstimate);
+    }
```

Comment:
This might give 'wrong results', because multiple robots might collide with the ball trajectory. It would be better to check for the closest one.